python/src/array.cpp

```diff
@@ -774,6 +774,25 @@ void init_array(py::module_& m) {
             return a.shape(0);
           })
       .def("__iter__", [](const array& a) { return ArrayPythonIterator(a); })
+      .def(py::pickle(
+          [](array& a) { // __getstate__
+            return py::make_tuple(
+                dtype_to_array_protocol(a.dtype()), tolist(a));
```

Comment:
I'm wondering about the use of `tolist` here. That will be a pretty inefficient way to pickle and un-pickle large arrays..

Did you consider any alternatives?

Reply:
Thanks for the benchmarks that's really interesting.

> For sure I can check how they are doing that in Numpy and take inspiration. 

Very good idea.. 

Another possibility is to convert the array to numpy and pickle that? I think this is what Jax does: https://github.com/google/jax/discussions/8494#discussioncomment-1619664


> PS: I've benchmarked also deepcopy(), and this such fast that I'm worried that is not actually behaving as expected!

This is expected given the way you implemented it as it is actually a shallow copy. But that's actually ok as arrays are "immutable" hence it doesn't really make sense to have a true deep copy.